WordPress/openverse
# Add low-level TS API client with generated types

## Description


Based on my work exploring this concept in https://git.sr.ht/~sara/openverse-api-client.

There are major differences from that version:
1. I have only included the TypeScript client in this PR. (Previous versions included a Python client).
2. The TypeScript client uses [`openapi-fetch`](https://openapi-ts.dev/openapi-fetch/) and it's TypeScript support package to generate TypeScript interfaces from the OpenAPI schema, instead of generating them with Python and Jinja2 templates.

I've already demonstrated the value of the JavaScript client for even our own frontend. The Python client, on the other hand, would be more useful for things like the Openverse Slack Reaction app (which isn't part of the project yet). For that reason, I think it is necessary to scope this only to the JS client. It also makes this easier to review, because the JS client is substantially simpler than the Python client. That's thanks to the excellence of the `openapi-fetch` and `openapi-typescript` packages. Nothing of that quality exists for Python. We will need to do much more work for a generated Python client, and it deserves to be its own pull request.

To briefly document the approach:

- I've introduced one new package, `@openverse/api-client`. The package houses the generated types and exports a client wrapping those types.
- The package also implements an [`openapi-fetch` Middleware](https://openapi-ts.dev/openapi-fetch/middleware-auth) to handle the Openverse API authentication flow.
  - This flow is tested with `nock`. However, it might be worth adding another test suite of integration level tests that run against a "real" API instance (even if it's the local API). I think that would be something good for a follow-up PR, rather than in this one, as it will require additional tooling to make possible in CI/CD.
  - For an example of the type of tooling I'm talking about, [see this commit which removes similar tooling used to enable the previously included Python package to run tests against the local dev API instance](https://github.com/WordPress/openverse/pull/4538/commits/fb393ee83e8b9e6d208e644c6c9f66a8584df092). It is possible, and even desirable, to have this. But it should be in a separate PR to ensure ease of review.
- Package publication is _not_ included, but the package's `build` script already includes checks to ensure the package builds and passes basic checks required for publication. A separate PR should follow to document and implement a manual publication process. We can turn that into an automated process later on.

Further improvements can be made to create higher level APIs that build upon this "low-level" version, as discussed by @obulat (who raised the question in the first place) and myself in the comments of this PR. Additionally, we could consider creating an intermediate package `@openverse/types` to house the generated OpenAPI schema types, as well as further support types like we use in the frontend's `src/types` directory, such as definitions of media types, and so forth. That package could also export specific components from `components["schemas"]`, like a `type Media = components["schemas"].Audio | components["schemas"].Image`, and so forth.

## Testing Instructions

Review the documentation for the package (and the information above) to understand the goals and approach. Then review the code. Due to the nature of the `openapi-fetch` package, there is actually relatively little code to review in this PR. (This was not the case in previous versions which also included a Python version.) Feel free to review the generated code to get a feeling for how it looks.

I've left comments on some changes in the PR that I hope will clarify things I anticipated may have questions asked about. Please feel free to ask for further clarification in any case, of course!

The main code to review carefully is the authentication middleware. This middleware is based on the strategy proven in our own frontend application, [by the `api-token` Nuxt plugin](https://github.com/WordPress/openverse/blob/faf599ba7d1c1973e8e062a19f0b5989588ee13e/frontend/src/plugins/api-token.server.ts). As you read the middleware code, review the test cases for it as well, in the tests included in the package.

You can also check out the branch and try the client out in the unit tests, to see how it feels to work with. Nothing is set in stone, and it is only the "low-level" version. As Olga and I have discussed in the comments on this PR, it's worth considering a future where we have higher level functions like `client.image.search`, and so on, which may have better ergonomics and discoverability than `client.GET("/v1/images/", { params: { query: { q } } })`.

We should aim to use the package inside our frontend (and elsewhere, if appropriate) as much as possible to exercise it and understand what kinds of higher level wrapper functions would be beneficial.

## Checklist


- [N/A] I ran the DAG documentation generator (`./ov just catalog/generate-docs` for catalog
      PRs) or the media properties generator (`./ov just catalog/generate-docs media-props`
      for the catalog or `./ov just api/generate-docs` for the API) where applicable.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
